controller/share_manager_controller.go

```diff
@@ -1050,6 +1050,69 @@ func (c *ShareManagerController) getShareManagerTolerationsFromStorageClass(sc *
 	return tolerations
 }
 
+func (c *ShareManagerController) checkStorageNetworkApplied() (bool, error) {
+	targetSettings := []types.SettingName{types.SettingNameStorageNetwork, types.SettingNameStorageNetworkForRWXVolumeEnabled}
+	for _, item := range targetSettings {
+		if applied, err := c.ds.GetSettingApplied(item); err != nil || !applied {
+			return applied, err
+		}
+	}
+	return true, nil
+}
+
+func (c *ShareManagerController) cleanupServiceAndEndpoint(shareManager *longhorn.ShareManager) error {
+	service, err := c.ds.GetService(c.namespace, shareManager.Name)
+	if err != nil {
+		// if NotFound, means the service/endpoint is already cleaned up
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return errors.Wrapf(err, "failed to get service for share manager %v", shareManager.Name)
+	}
+
+	applied, err := c.checkStorageNetworkApplied()
+	if err != nil {
+		return errors.Wrapf(err, "failed to check if the storage network setting is applied")
+	}
+	if !applied {
+		c.logger.Warnf("`StorageNetwork` related settings are not applied, do nothing.")
+		return nil
+	}
+
+	storageNetwork, err := c.ds.GetSettingWithAutoFillingRO(types.SettingNameStorageNetwork)
```

Comment:
Nit: I am thinking if we should move the service and endpoint check validation to a separate method like what you've done for `c.checkStorageNetworkApplied()`. I am also fine with the current, what do you think?

Reply:
Hi @c3y1huang, I wrapped some helpful functions like `validateStorageNetworkForRWXVolume` and `canCleanupServiceAndEndpoint`.

Please help to recheck it. Thanks!